### PR TITLE
[PROTOCOL-551] Teller Protocol - Poloygon 

### PR DESCRIPTION
**Pull request by sypeer**
It is:
- 0xPolygon's version of the Teller protocol
- Based on the Diamonds contract structure https://eips.ethereum.org/EIPS/eip-2535

### CODE REVIEW

File: types/custom/config-types.d.ts [hunk 1]
```diff
@@ -99,7 +99,7 @@ export interface Signers extends Array<Address> {}
 
 export interface NetworkTokens {
   erc20: Tokens
-  compound: Tokens
+  compound: Tokens // TODO: Rename/refactor for agnostic strategy token/tokens
```
Comment (passabilities) [inline types/custom/config-types.d.ts:102]: can just add an `aave` field

**Review by passabilities — CHANGES_REQUESTED** [review]

File: config/tokens.ts [hunk 2]
```diff
@@ -97,6 +97,50 @@ export const tokens: Record<string, NetworkTokens> = {
       LINK: '0x20fE562d797A42Dcb3399062AE9546cd06f63280',
     },
   },
+  polygon: {
+    compound: {
+      ADAI: '0x27F8D03b3a2196956ED754baDc28D73be8830A6e',
+      AUSDC: '0x1a13F4Ca1d028320A707D99520AbFefca3998b7F',
+      AETH: '0x28424507fefb6f7f8E9D3860F56504E4e5f5f390', //amWETH
+    },
```
[the remark this answers is not in the document]
Comment (sypeer) [reply config/tokens.ts:105]: Agreed, need to make a few changes to the deployment

File: config/signers.ts [hunk 3]
```diff
@@ -19,4 +19,6 @@ export const signers: Record<string, Signers> = {
   hardhat: [],
   localhost: [],
   mainnet: [],
+  polygon: [],
```
[the remark this answers is not in the document]
Comment (sypeer) [reply config/signers.ts:22]: yes sir, getting it up once the nodes are finalized

File: contracts/lending/ttoken/strategies/aave/TTokenAaveStrategy_1.sol [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    ) internal {
+        // Calculate amount to rebalance
+        uint16 medianRatio =
+            (aaveStore().balanceRatioMax + aaveStore().balanceRatioMin) / 2;
+        uint256 requiredBal =
+            NumbersLib.percent(
+                storedBal + aaveBal - amount,
+                medianRatio - storedRatio
+            );
+        uint256 redeemAmount = amount - storedBal + requiredBal;
```
[the remark this answers is not in the document]
Comment (sypeer) [reply contracts/lending/ttoken/strategies/aave/TTokenAaveStrategy_1.sol]: Done

File: contracts/market/collateral/CollateralEscrow_V1.sol [hunk 5]
```diff
@@ -18,12 +18,12 @@ import { Address } from "@openzeppelin/contracts/utils/Address.sol";
 contract CollateralEscrow_V1 is OwnableUpgradeable, ICollateralEscrow {
     mapping(uint256 => uint256) internal _balances; // loanID -> collateral
     address public collateralToken;
-    bool internal _isWETH;
+    bool internal _isWETH; // Should this be renamed to _isNetworkToken? more agnostic in cases of WMATIC
```
Comment (passabilities) [inline contracts/market/collateral/CollateralEscrow_V1.sol:21]: `_isWrappedNativeToken`

File: contracts/market/libraries/LibCollateral.sol [hunk 6]
```diff
@@ -97,8 +97,8 @@ library LibCollateral {
                 AppStorageLib.store().collateralEscrowBeacon.cloneProxy("");
             ICollateralEscrow(escrow).init(
                 token,
-                // Check if collateral token is WETH
-                token == AppStorageLib.store().assetAddresses["WETH"]
+                // Check if collateral token is the base network token
+                token == AppStorageLib.store().baseNetworkToken
```
Comment (passabilities) [inline contracts/market/libraries/LibCollateral.sol:101]: `wrappedNativeToken`

**Review by passabilities — CHANGES_REQUESTED** [review]

File: config/chainlink.ts [hunk 7]
```diff
@@ -121,6 +121,50 @@ export const chainlink: Record<string, Chainlink> = {
...
+      baseTokenName: 'LINK',
+      quoteTokenName: 'DAI',
+      address: '0xd9FFdb71EbE7496cC440152d43986Aae0AB76665', // LINK-USD
+    },
+    LINK_USDC: {
+      baseTokenName: 'LINK',
+      quoteTokenName: 'USDC',
+      address: '0xd9FFdb71EbE7496cC440152d43986Aae0AB76665', // LINK-USD
+    },
+  },
```
[the remark this answers is not in the document]
Comment (sypeer) [reply config/chainlink.ts:145]: Added a few, would you like to add support for ada or doge? &#x1f914;

[the remark this answers is not in the document]
Comment (passabilities) [reply config/chainlink.ts]: most definitely NOT

**Review by dumebi — DISMISSED** [review]

File: contracts/escrow/dapps/libraries/LibDapps.sol [hunk 8]
```diff
@@ -35,7 +37,7 @@ library LibDapps {
         return
             IAaveLendingPool(
                 IAaveLendingPoolAddressesProvider(
-                    0xB53C1a33016B2DC2fF3653530bfF1848a515c8c5
+                    0xd05e3E715d945B59290df0ae8eF85c1BdB684744 // Polygon mainnet
```
[the remark this answers is not in the document]
Comment (sypeer) [reply contracts/escrow/dapps/libraries/LibDapps.sol:40]: will update with the dapp router/registry contract

File: contracts/escrow/dapps/libraries/LibSushiswap.sol [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+} from "../../../shared/interfaces/IUniswapV2Router.sol";
+
+// Libraries
+import {
+    SafeERC20
+} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+library LibSushiswap {
+    IUniswapV2Router constant ROUTER =
+        IUniswapV2Router(0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506); // Sushiswap polygon
```
[the remark this answers is not in the document]
Comment (sypeer) [reply contracts/escrow/dapps/libraries/LibSushiswap.sol:17]: same

**Review by dumebi — DISMISSED** [review]

File: config/signers.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (passabilities) [reply config/signers.ts:22]: unless zk is working before we launch, don't forget this

File: config/tokens.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (passabilities) [reply config/tokens.ts:105]: ?

File: contracts/escrow/dapps/PoolTogetherFacet.sol [hunk 10]
```diff
@@ -68,18 +66,17 @@ contract PoolTogetherFacet is PausableMods, DappMods {
...
+        // Set token allowance to Prize Pool
+        LibEscrow.e(loanID).setTokenAllowance(tokenAddress, address(prizePool));
+        bytes memory callData = abi.encodeWithSelector(
+            PrizePoolInterface.depositTo.selector,
+            address(LibEscrow.e(loanID)), // Tickets sent to Escrow
+            amount,
+            ticketAddress,
+            address(this) // Referrer is the Teller Diamond
+        );
+        LibDapps.s().loanEscrows[loanID].callDapp(address(prizePool), callData);
```
Comment (passabilities) [inline contracts/escrow/dapps/PoolTogetherFacet.sol:79]: should use `LibEscrow.e(loanID)` just to stay consistent

File: contracts/escrow/dapps/PoolTogetherFacet.sol [hunk 11]
```diff
@@ -111,24 +108,13 @@ contract PoolTogetherFacet is PausableMods, DappMods {
...
-                maxExitFee
-            );
+        bytes memory callData = abi.encodeWithSelector(
+            PrizePoolInterface.withdrawInstantlyFrom.selector,
+            address(LibEscrow.e(loanID)),
+            amount,
+            ticketAddress,
+            0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF // max exit fee
+        );
         LibDapps.s().loanEscrows[loanID].callDapp(address(prizePool), callData);
```
Comment (passabilities) [inline contracts/escrow/dapps/PoolTogetherFacet.sol:118]: same

File: contracts/escrow/dapps/libraries/LibDapps.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (passabilities) [reply contracts/escrow/dapps/libraries/LibDapps.sol:40]: is this hard coded for mumbai or poly mainnet?

File: contracts/escrow/dapps/libraries/LibSushiswap.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (passabilities) [reply contracts/escrow/dapps/libraries/LibSushiswap.sol:17]: is this hard coded for mumbai or poly mainnet?

File: contracts/market/collateral/CollateralEscrow_V1.sol [hunk 12]
```diff
@@ -18,22 +18,25 @@ import { Address } from "@openzeppelin/contracts/utils/Address.sol";
...
     mapping(uint256 => uint256) internal _balances; // loanID -> collateral
     address public collateralToken;
-    bool internal _isWETH;
+    bool internal _isWrappedNativeToken; // Should this be renamed to _isNetworkToken? more agnostic in cases of WMATIC
 
     /**
      * @notice it initializes an escrow
      * @param tokenAddress the address of the collateral token to be stored
-     * @param isWETH check if it's wrapped Ethereum
+     * @param isWrappedNativeToken check if it's wrapped Ethereum/MATIC
```
Comment (passabilities) [inline contracts/market/collateral/CollateralEscrow_V1.sol:26]: how about
> boolean indicating if its the wrapped native token

File: contracts/settings/asset/AssetSettingsDataFacet.sol [hunk 13]
```diff
@@ -7,6 +7,8 @@ import { ADMIN, AUTHORIZED } from "../../shared/roles.sol";
 
 // Interfaces
 import { ICErc20 } from "../../shared/interfaces/ICErc20.sol";
+import { IAToken } from "../../shared/interfaces/IAToken.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
```
Comment (passabilities) [inline contracts/settings/asset/AssetSettingsDataFacet.sol:11]: to stay consistent, lets import the named interface

File: deploy/protocol.ts [hunk 14]
```diff
@@ -27,6 +27,11 @@ const deployProtocol: DeployFunction = async (hre) => {
   const loansEscrowBeacon = await deployLoansEscrowBeacon(hre)
   const collateralEscrowBeacon = await deployCollateralEscrowBeacon(hre)
   const tTokenBeacon = await deployTTokenBeacon(hre)
+  const nftDictionary = await contracts.get('TellerNFTDictionary')
+
+  const tokens = getTokens(network)
+  const wrappedNativeToken =
+    networkName == 'polygon_mumbai' ? tokens.erc20.WMATIC : tokens.erc20.WETH
```
Comment (passabilities) [inline deploy/protocol.ts:34]: will want this to also include `polygon` mainnet network as well

File: tasks/fork.ts [hunk 15]
```diff
@@ -86,15 +86,23 @@ subtask('fork:fund-deployer').setAction(async (args, hre) => {
...
+        amount: hre.ethers.utils.parseEther('1000'),
+        hre,
+      }),
+      getFunds({
+        to: deployer,
+        tokenSym: 'MATIC',
+        amount: hre.ethers.utils.parseEther('10000'),
+        hre,
+      }),
+    ])
```
Comment (passabilities) [inline tasks/fork.ts:106]: should do a network check and only get the main native asset

File: tasks/fork.ts [hunk 16]
```diff
@@ -86,15 +86,23 @@ subtask('fork:fund-deployer').setAction(async (args, hre) => {
   const { ethers } = hre
 
   const [mainAccount] = await hre.getUnnamedAccounts()
-  const { deployer } = await hre.getNamedAccounts()
+  const { deployer, funder } = await hre.getNamedAccounts()
```
Comment (passabilities) [inline tasks/fork.ts:89]: is `funder` being used?

File: test/integration/escrow/dapps/compound.test.ts [hunk 17]
```diff
@@ -101,3 +101,104 @@
 //     })
 //   }
 // })
+import { getMarkets } from '../../../../config'
+import { getPlatformSetting } from '../../../../tasks'
+import { Market } from '../../../../types/custom/config-types'
+import { ERC20, ICErc20, ITellerDiamond } from '../../../../types/typechain'
```
Comment (passabilities) [inline test/integration/escrow/dapps/compound.test.ts:107]: can remove the commented out tests?

File: test/integration/escrow/dapps/uniswap.test.ts [hunk 18]
```diff
@@ -129,3 +129,133 @@
 //     })
 //   }
 // })
+import { getMarkets } from '../../../../config'
+import { getPlatformSetting } from '../../../../tasks'
+import { Market } from '../../../../types/custom/config-types'
+import { ERC20, ITellerDiamond } from '../../../../types/typechain'
```
Comment (passabilities) [inline test/integration/escrow/dapps/uniswap.test.ts:135]: can remove the commented out tests?

File: config/signers.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sypeer) [reply config/signers.ts:22]: Added

File: contracts/escrow/dapps/libraries/LibDapps.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (sypeer) [reply contracts/escrow/dapps/libraries/LibDapps.sol:40]: poly mainnet here

File: contracts/escrow/dapps/libraries/LibSushiswap.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (sypeer) [reply contracts/escrow/dapps/libraries/LibSushiswap.sol:17]: poly mainnet

File: deploy/protocol.ts (hunk 14 above)
Comment (sypeer) [reply deploy/protocol.ts:34]: changed to mainnet since Mumbai isn't feasible

File: test/integration/escrow/dapps/uniswap.test.ts (hunk 18 above)
Comment (sypeer) [reply test/integration/escrow/dapps/uniswap.test.ts:135]: removed

File: test/integration/escrow/dapps/compound.test.ts (hunk 17 above)
Comment (sypeer) [reply test/integration/escrow/dapps/compound.test.ts:107]: removed

File: config/asset-settings.ts [hunk 19]
```diff
@@ -9,19 +9,31 @@ const mainnetAssetSettings: AssetSettings = {
     { key: 'pPool', value: 'PCDAI', type: AssetType.Address },
     { key: 'MaxLoanAmount', value: 25000, type: AssetType.Amount },
     { key: 'MaxTVL', value: 10000000, type: AssetType.Amount },
-    { key: 'MaxDebtRatio', value: 5000, type: AssetType.Uint },
+    { key: 'MaxDebtRatio', value: 10000, type: AssetType.Uint },
```
[the remark this answers is not in the document]
Comment (sypeer) [reply config/asset-settings.ts:12]: changed

File: config/asset-settings.ts [hunk 20]
```diff
@@ -9,19 +9,31 @@ const mainnetAssetSettings: AssetSettings = {
...
+}
+
 export const assetSettings: Record<string, AssetSettings> = {
   kovan: {
     DAI: [
       { key: 'cToken', value: 'CDAI', type: AssetType.Address },
       { key: 'MaxLoanAmount', value: 25000, type: AssetType.Amount },
       { key: 'MaxTVL', value: 100000, type: AssetType.Amount },
-      { key: 'MaxDebtRatio', value: 5000, type: AssetType.Uint },
+      { key: 'MaxDebtRatio', value: 10000, type: AssetType.Uint },
```
[the remark this answers is not in the document]
Comment (sypeer) [reply config/asset-settings.ts:36]: same

File: config/asset-settings.ts [hunk 21]
```diff
@@ -42,12 +54,14 @@ export const assetSettings: Record<string, AssetSettings> = {
...
```
[the remark this answers is not in the document]
Comment (sypeer) [reply config/asset-settings.ts:57]: done

File: config/asset-settings.ts [hunk 22]
```diff
@@ -32,7 +44,7 @@ export const assetSettings: Record<string, AssetSettings> = {
       { key: 'pPool', value: 'PCDAI', type: AssetType.Address },
       { key: 'MaxLoanAmount', value: 25000, type: AssetType.Amount },
       { key: 'MaxTVL', value: 100000, type: AssetType.Amount },
-      { key: 'MaxDebtRatio', value: 5000, type: AssetType.Uint },
+      { key: 'MaxDebtRatio', value: 10000, type: AssetType.Uint },
```
[the remark this answers is not in the document]
Comment (sypeer) [reply config/asset-settings.ts]: si senior

File: tasks/fork.ts (hunk 16 above)
Comment (sypeer) [reply tasks/fork.ts:89]: Doesn't look like it, removed

File: tasks/fork.ts (hunk 15 above)
Comment (sypeer) [reply tasks/fork.ts:106]: Added in

File: config/tokens.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sypeer) [reply config/tokens.ts:105]: changed `compound` to `underlyingToken` instead of adding `aave`, keeps it more agnostic imo

File: tasks/fork.ts [hunk 23]
```diff
@@ -83,18 +83,27 @@ task('fork', 'Forks a chain and starts a JSON-RPC server of that forked chain')
...
   if (
     ethers.utils.getAddress(mainAccount) !== ethers.utils.getAddress(deployer)
   ) {
-    await getFunds({
-      to: deployer,
-      tokenSym: 'ETH',
-      amount: hre.ethers.utils.parseEther('1000'),
-      hre,
-    })
+    if (network.name !== 'polygon') {
```
Comment (passabilities) [inline tasks/fork.ts:93]: I think you meant `===` but also don't forget about **mumbai**

File: contracts/escrow/dapps/PoolTogetherFacet.sol (hunk 10 above)
Comment (mkhoury93) [reply contracts/escrow/dapps/PoolTogetherFacet.sol:79]: I like `escrow` more ;)

**Review by mkhoury93 — APPROVED** [review]

File: tasks/fork.ts (hunk 23 above)
Comment (sypeer) [reply tasks/fork.ts]: fixed

**Review by mkhoury93 — APPROVED** [review]

**Review by passabilities — APPROVED** [review]

**Pull request merged**